src/clj/rems/db/duo.clj

```diff
@@ -0,0 +1,144 @@
+(ns rems.db.duo
+  (:require [clj-http.client :as http]
+            [clojure.data.csv :as csv]
+            [clojure.java.io :as io]
+            [clojure.set :refer [difference]]
+            [clojure.string :as str]
+            [medley.core :refer [remove-vals update-existing]]
+            [rems.config]
+            [rems.db.core :as db]
+            [rems.json :as json]
+            [rems.schema-base :refer [LocalizedString]]
+            [schema.coerce :as coerce]
+            [schema.core :as s]))
+
+;; The CSV format has English texts only
+(s/defschema DuoCodeCsv
+  {:id s/Str
+   (s/optional-key :shorthand) (s/maybe s/Str)
+   :label s/Str
+   :description s/Str})
+
+;; The DB format has localized texts
+(s/defschema DuoCodeDb
+  {:id s/Str
+   (s/optional-key :shorthand) (s/maybe s/Str)
+   :label LocalizedString
+   :description LocalizedString})
+
+(def ^:private coerce-DuoCodeCsv
+  (coerce/coercer! DuoCodeCsv coerce/string-coercion-matcher))
+
+(def ^:private coerce-DuoCodeDb
+  (coerce/coercer! DuoCodeDb coerce/string-coercion-matcher))
+
+(def ^:private validate-DuoCodeDb
+  (s/validator DuoCodeDb))
+
+(defn- rows->maps
+  "Transform CSV `rows` into maps with the keys from the first header row."
+  [rows]
+  (mapv zipmap
+        (->> (first rows) ; header
+             (map keyword)
+             repeat)
+        (rest rows)))
+
+(defn upsert-duo-code!
+  "Insert a new or update an existing DUO code into the database."
+  [duo]
+  (let [data (-> duo
+                 validate-DuoCodeDb
+                 (dissoc :id)
+                 (->> (remove-vals #(cond (string? %) (str/blank? %)
+                                          (coll? %) (empty? %)
+                                          :else %)))
+                 json/generate-string)]
+    (when (:id duo)
+      (db/upsert-duo-code! {:id (:id duo)
+                            :data data}))))
+
+(defn- read-zip-entries
+  "Read the zip-file entries from the `stream` and returns those matching `re`."
+  [stream re]
+  (when stream
+    (loop [files []]
+      (if-let [entry (.getNextEntry stream)]
+        (if (re-matches re (.getName entry))
+          (let [baos (java.io.ByteArrayOutputStream.)]
+            (io/copy stream baos)
+            (recur (conj files {:name (.getName entry)
+                                :bytes (.toByteArray baos)})))
+          (recur files))
+        files))))
+
+(defn fetch-duo-releases-latest
+  "Fetches the latest DUO codes from the GitHub repository.
+
+  Returns the unpacked, formatted result."
+  []
+  (let [release-response (http/get "https://api.github.com/repos/EBISPOT/DUO/releases/latest"
```

Comment:
do we plan to utilize versioning in some capacity? eg. `:tag`. or in other words, that what the game plan is now regarding them :)

Reply:
Well my idea is that at some point we need to say that the version has changed. Since we have no operational monitoring capability in REMS itself, and this process is now modeled as a manual process (done by developers), we could perhaps just store the version number and alert the developer if it has changed when they check the new codes.

Reply:
ok! so in other words fetching latest DUO codes could evolve into the direction of a CLI command? maybe we can help the administrators with some tag fetching automation, if necessary

Reply:
I think a CLI command won't be very useful here, and it's not what I'm suggesting. If any new type comes along, we must implement features to ask the right follow-up questions or handle the compatibility check correctly? So I'm suggesting we developers follow up on the releases with the commented code and make any necessary changes.